src/stream_processor/flb_sp_snapshot.c

```diff
@@ -0,0 +1,275 @@
+    msgpack_object *obj;
+
+    ok = MSGPACK_UNPACK_SUCCESS;
+    msgpack_unpacked_init(&result);
+
+    if (buf_size <= 0) {
+        return -1;
+    }
+
+    snapshot = (struct flb_sp_snapshot *) task->snapshot;
+
+    /* Create a snapshot pgae if the list is empty */
+    if (mk_list_is_empty(&snapshot->pages) == 0) {
+        page = snapshot_page_create();
+        if (!page) {
+            flb_errno();
+            return -1;
+        }
+
+        mk_list_add(&page->_head, &snapshot->pages);
+    }
+    else {
+        page = mk_list_entry_last(&snapshot->pages, struct flb_sp_snapshot_page, _head);
+
+        if (snapshot_page_is_full(page, buf_size)) {
+            page = snapshot_page_create();
+            if (!page) {
+                flb_errno();
+                return -1;
+            }
+
+            mk_list_add(&page->_head, &snapshot->pages);
+        }
+    }
+
+    memcpy(page->snapshot_page + page->end_pos, buf_data, buf_size);
+    page->end_pos = page->end_pos + buf_size;
+
+    /* Get the last timestamp */
+    while (msgpack_unpack_next(&result, page->snapshot_page,
+                               page->end_pos - page->start_pos, &off) == ok) {
+        flb_time_pop_from_msgpack(&tm, &result, &obj);
+    }
+
+    msgpack_unpacked_destroy(&result);
+
+    snapshot->records++;
+    snapshot->size = snapshot->size + buf_size;
+
+    /* Remove records from snapshot pages based on time/length window */
+    snapshot_cleanup(snapshot, tms);
+
+    return 0;
+}
+
+int flb_sp_snapshot_flush(struct flb_sp *sp, struct flb_sp_task *task,
+                          char **out_buf_data, size_t *out_buf_size)
+{
+    size_t off;
+    size_t page_size;
+    char *snapshot_name;
+    struct flb_sp_cmd *cmd;
+    struct mk_list *tmp;
+    struct mk_list *head;
+    struct mk_list *snapshot_head;
+    struct flb_sp_task *snapshot_task;
+    struct flb_sp_snapshot *snapshot;
+    struct flb_sp_snapshot_page *page;
+
+    off = 0;
+    cmd = task->cmd;
+    snapshot_name = flb_sp_snapshot_name_from_flush(cmd->stream_name);
+
+    /* Lookup Tasks that matches the incoming instance data */
+    mk_list_foreach(head, &sp->tasks) {
+        snapshot_task = mk_list_entry(head, struct flb_sp_task, _head);
+        cmd = snapshot_task->cmd;
+
+        if (cmd->type == FLB_SP_CREATE_SNAPSHOT &&
+            flb_sds_cmp(cmd->stream_name, snapshot_name,
+                        strlen(snapshot_name)) == 0) {
+
+            snapshot = (struct flb_sp_snapshot *) snapshot_task->snapshot;
+
+            if (snapshot->size == 0) {
+                break;
+            }
+
+            if (*out_buf_data == NULL) {
+                *out_buf_data = (char *) flb_malloc(snapshot->size);
+                if (!*out_buf_data) {
+                    flb_errno();
+                    return -1;
+                }
+                *out_buf_size = snapshot->size;
+            }
+            else {
+                *out_buf_data = (char *) flb_realloc(*out_buf_data,
```

Comment:
here please use a temporal variable. Otherwise if the call fails it will return null, and the original memory area will be lost.